### ISSUE TITLE
Darktable 5.0 <-> GIMP 3.0 interface for windows

### ISSUE THREAD

**Issue by jenshannoschwalm** [darktable-org/darktable#17853]
### Describe the bug

It seems that on windows darktable master is still not recognized in GIMP 3.0 as a plugin so developing a raw via darktable is broken.

### Steps to reproduce

Can't test as not a windows user and not having access to a windows system at all.

### Expected behavior

Interface should work

### Where did you obtain darktable from?

self compiled

### darktable version

current master

### What OS are you using?

Windows

### What is the version of your OS?

???

### Are you using OpenCL GPU in darktable?

None

### Please provide additional context if applicable. You can attach files too, but might need to rename to .txt or .zip

See as reference

https://gitlab.gnome.org/GNOME/gimp/-/issues/12260
https://discuss.pixls.us/t/gimp-darktable-not-working-with-gimp-3-0-0-rc1-and-current-dt-from-master/46609/12

Opening this to keep track

**Comment by jenshannoschwalm** [comment]
@zisoft would you be able to check dt master vs gimp 3.0rc on windows so we might get it fixed or confirmed it works with dt 5.0?

**Comment by zisoft** [comment]
I'll try my very best :-)

This is my PC at work and I have no admin privileges here, so neither darktable nor gimp are installed in standard windows paths.

Can you tell me what steps are needed to test this?

**Comment by jenshannoschwalm** [comment]
GIMP with latest 3.0RC and dt master should be installed.

1. Starting gimp should show dt as a plugin
2. loading a raw via dt should work :-)

**Comment by zisoft** [comment]
I am afraid I cannot help here much.
Without admin privileges I cannot install in Windows default paths so gimp cannot find my self-compiled darktable.
Is there any way to tell gimp where to look?

**Comment by dtrtuser** [comment]
Just FYI: tried on 2 different computers (windows 10 and windows 11). Gimp 3.0.0-RC1 installed in default location (C:\Program Files\GIMP 3) and darktable from current master also installed in default location (C:\Program Files\darktable) and Gimp can't find darktable.

**Comment by jenshannoschwalm** [comment]
@dtrtuser i have absolutely no idea about windows and what this means "can't find darktable". I think there might be some "registry" involved. 

@Jehan maybe you could step in here as we definitely want to get this working? I checked again the protocol part on dt side and that seems to be good. Any idea? I looked into gimp  `plug-ins/file-raw/file-darktable.c` and couldn't get what might be wrong on our or GIMP's side.

**Comment by dtrtuser** [comment]
@jenshannoschwalm "can't find darktable" means that if in Gimp I go to "Edit" -> "Preferences" -> "Image Import & Export" darktable is not listed in the plugins.

**Comment by jenshannoschwalm** [comment]
Could you check if the dark table.exe file exists at the tested location?

**Comment by Jehan** [comment]
@jenshannoschwalm 
> @Jehan maybe you could step in here as we definitely want to get this working? 

So looking in the code, what we do on Windows to find the raw developer executable is that we look for the value of a [given registry key](https://gitlab.gnome.org/GNOME/gimp/-/blob/e38dae1bb116f8a8fe752a555ce546314e8fa6ff/plug-ins/file-raw/file-raw-utils.c#L132-137), checking it first on `HKEY_CURRENT_USER`, then on `HKEY_LOCAL_MACHINE` (which I understand as being either user-installed or system-installed).

For darktable in particular, this [key](https://gitlab.gnome.org/GNOME/gimp/-/blob/e38dae1bb116f8a8fe752a555ce546314e8fa6ff/plug-ins/file-raw/file-darktable.c#L35) is `SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths\darktable` (maybe adding a `.exe` if I understand correctly the code in `file_raw_get_executable_path()` doing a concat().

Maybe this key which you used to set in the Windows registry has been changed?

**Comment by jenshannoschwalm** [comment]
We will find out :-)

**Comment by dtrtuser** [comment]
> Could you check if the dark table.exe file exists at the tested location?

Yes, it’s there.

**Comment by dtrtuser** [comment]
I made some progress (but still not working) ...

It seems that **Gimp 3RC1 package for Windows is missing 2 files** in order to be able to spawn darktable (g_spawn_sync):

- gspawn-win64-helper-console.exe
- gspawn-win64-helper.exe

If I copy these 2 files in \Program Files\Gimp 3\bin, then Gimp lists darktable in the preferences and uses darktable to open raw files. I can do my edits in darktable but when I close it (and expect to get the image in Gimp), Gimp pops up an error message:

```
Calling error for procedure 'gimp-file-load':
Error opening file 'C:\Users\dtrtuser\AppData\Local\Temp\darktable_YLL3X2\image.exr' for reading
Opening 'D:\Backup Pictures\2024\Imported on Wednesday, November 27, 2024\20241127-_7RV1846.ARW' failed:
(NULL) plug-in could not open image

```

If I look in the C:\Users\dtrtuser\AppData\Local\Temp\darktable_YLL3X2\ folder, it's empty.

Also below is the output of Gimp when running from command line:

```
## stdout ##
     0.0001 [dt starting]
 C:\Program Files\darktable\bin\darktable.exe --gimp file D:\Backup Pictures\2024\Imported on Wednesday, November 27, 2024\20241127-_7RV1850.ARW
    11.6218 [export_job] exported to `C:UsersdtrtuserAppDataLocalTempdarktable_YLL3X2/image.exr'
<<<gimp
C:\Users\dtrtuser\AppData\Local\Temp\darktable_YLL3X2/image.exr
gimp>>>

```

Please note that the path is missing the \ (backslash). Not sure if it is a log formatting issue or if darktable is really trying to output to this location (in which case it would explain why not file is truly created).

Hope this helps.

**Comment by Jehan** [comment]
> It seems that Gimp 3RC1 package for Windows is missing 2 files in order to be able to spawn darktable (g_spawn_sync):

Ah right. We tried to have our package do a small diet and over-deleted some files. This got fixed here for RC2: https://gitlab.gnome.org/GNOME/gimp/-/issues/12288

Though it looks like we only include `gspawn-win64-helper-console.exe` and not `gspawn-win64-helper.exe`.

@dtrtuser Could you verify if `gspawn-win64-helper.exe` is really useful or if only `gspawn-win64-helper-console.exe` is enough please? If the former, I guess we need an additional fix. If the latter, we should already good to go in GIMP side.

(also pinging @brunvonlope)

> For instance, rather than writing the image.exr file in C:\Users\dtrtuser\AppData\Local\Temp\darktable_YLL3X2\ the file is written instead in C:\Users\dtrtuser\UsersdtrtuserAppDataLocalTempdarktable_YLL3X2

For this part of the issue, it indeed feels like it's on darktable side.

Thanks for all the investigating/testing @dtrtuser !

**Comment by jenshannoschwalm** [comment]
> (Just to be clear: the output directory is properly created, but the image file (image.exr) is place in another one).

Ok, i'll do the homework here :-)

**Comment by brunvonlope** [comment]
> I removed gspawn-win64-helper.exe from \Program Files\Gimp 3\bin and it behaves the same, so this one might not be needed

Since you're taking the gspawn executable from some msystem prefix /bin dir there is a risk of glib and its dependencies version mismatching.

I would like to trigger two GIMP master builds to test the gspawn theory: one not touched and another with the additional gspawn. But this can time time since GNOME GitLab is completely broken.

**Comment by brunvonlope** [comment]
Also, I h've seen some comments about admin rights. I think they are not (and should not) be needed. I remember of using darktable without problems in my user space (before the API breakages begun).

**Comment by jenshannoschwalm** [comment]
@dtrtuser would you be able to test #17931 ? Indeed the file seperator was plain wrong on windows, i think it should be fine now.

**Comment by jenshannoschwalm** [comment]
Let's keep this open to discuss until all issues are *confirmed* to be fixed.

**Comment by dtrtuser** [comment]
> @dtrtuser would you be able to test #17931 ? Indeed the file seperator was plain wrong on windows, i think it should be fine now.

I am await from my computer and I won’t be able to test until Wednesday evening (US Mountain Time).

**Comment by dtrtuser** [comment]
I was actually able to install GIMP and build dt from master (with changes from #17931, I checked the source) on my work laptop but unfortunately same issue: the path to image.exr is still missing all the backslashes.

**Comment by jenshannoschwalm** [comment]
Ok, can you confirm that the path printed in the protocol is **correct** but the written file is located elsewhere?

**Comment by dtrtuser** [comment]
> Ok, can you confirm that the path printed in the protocol is **correct** but the written file is located elsewhere?

That's correct: both dt and Gimp are supposed to write/read the file in C:\Users\dtrtuser\AppData\Local\Temp\darktable_A3YQX2\ but dt writes it in C:\Users\dtrtuser\ and the file name is UsersdtrtuserAppDataLocalTempdarktable_A3YQX2image.exr

**Comment by dtrtuser** [comment]
> @dtrtuser would you check #17945 please?

It works, thank you!!!
